core/sessions/orm.go

```diff
@@ -52,58 +56,90 @@ func NewORM(db *sqlx.DB, sessionDuration time.Duration, lggr logger.Logger) ORM
 	return &orm{db, sessionDuration, lggr.Named("SessionsORM")}
 }
 
-// FindUser will return the one API user, or an error.
-func (o *orm) FindUser() (User, error) {
-	return o.findUser()
+// FindUser will attempt to return an API user by email.
+func (o *orm) FindUser(email string) (User, error) {
+	return o.findUser(email)
 }
 
-func (o *orm) findUser() (user User, err error) {
-	sql := "SELECT * FROM users ORDER BY created_at desc LIMIT 1"
-	err = o.db.Get(&user, sql)
+// FindUserByAPIToken will attempt to return an API user by email.
+func (o *orm) FindUserByAPIToken(apiToken string) (user User, err error) {
+	sql := "SELECT * FROM users WHERE token_key = $1"
+	err = o.db.Get(&user, sql, apiToken)
 	return
 }
 
-// AuthorizedUserWithSession will return the one API user if the Session ID exists
-// and hasn't expired, and update session's LastUsed field.
+func (o *orm) findUser(email string) (user User, err error) {
+	sql := "SELECT * FROM users WHERE email = $1"
+	err = o.db.Get(&user, sql, email)
+	return
+}
+
+// ListUsers will load and return all user rows from the db.
+func (o *orm) ListUsers() (users []User, err error) {
+	sql := "SELECT * FROM users ORDER BY role ASC;"
+	err = o.db.Select(&users, sql)
+	return
+}
+
+// AuthorizedUserWithSession will return the API user associated with the Session ID if it
+// exists and hasn't expired, and update session's LastUsed field.
 func (o *orm) AuthorizedUserWithSession(sessionID string) (User, error) {
 	if len(sessionID) == 0 {
 		return User{}, errors.New("Session ID cannot be empty")
 	}
 
+	// First find user based on session token
+	var email *string
+	if err := o.db.Get(&email, "SELECT email FROM sessions WHERE id = $1", sessionID); err != nil {
+		return User{}, err
+	}
+	user, err := o.FindUser(*email)
+	if err != nil {
+		return User{}, err
+	}
+
+	// Sesssion valid and tied to user, update last_used
 	result, err := o.db.Exec("UPDATE sessions SET last_used = now() WHERE id = $1 AND last_used + $2 >= now()", sessionID, o.sessionDuration)
 	if err != nil {
 		return User{}, err
 	}
+
 	rowsAffected, err := result.RowsAffected()
 	if err != nil {
 		return User{}, err
 	}
 	if rowsAffected == 0 {
 		return User{}, sql.ErrNoRows
 	}
-	return o.FindUser()
+
+	return user, nil
 }
 
-// DeleteUser will delete the API User in the db.
-func (o *orm) DeleteUser() error {
+// DeleteUser will delete an API User and sessions by email.
+func (o *orm) DeleteUser(email string) error {
 	ctx, cancel := pg.DefaultQueryCtx()
 	defer cancel()
 	return pg.SqlxTransaction(ctx, o.db, o.lggr, func(tx pg.Queryer) error {
-		if _, err := tx.Exec("DELETE FROM users"); err != nil {
+		if _, err := tx.Exec("DELETE FROM sessions WHERE email = $1", email); err != nil {
 			return err
 		}
-
-		_, err := tx.Exec("DELETE FROM sessions")
+		_, err := tx.Exec("DELETE FROM users WHERE email = $1", email)
```

Comment:
Why not use a foreign key with cascade on sessions(email) then the delete is handled automatically

Reply:
Good idea, updated